File: examples/java/lifecycle/src/main/java/io/northrays/examples/Lifecycle.java
// Copyright Daytona Platforms Inc.
// SPDX-License-Identifier: Apache-2.0

package io.northrays.examples;

import io.northrays.sdk.Northrays;
import io.northrays.sdk.Sandbox;
import io.northrays.sdk.model.ListSandboxesQuery;
import io.northrays.sdk.model.SandboxListSortDirection;
import io.northrays.sdk.model.SandboxListSortField;
import io.northrays.sdk.model.SandboxState;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Lifecycle {
    public static void main(String[] args) {
        try (Northrays northrays = new Northrays()) {
            System.out.println("Creating sandbox");
            Sandbox sandbox = northrays.create();
            System.out.println("Sandbox created: " + sandbox.getId() + " (state: " + sandbox.getState() + ")");

            try {
                Map<String, String> labels = new HashMap<>();
                labels.put("test", "lifecycle");
                sandbox.setLabels(labels);
                System.out.println("Labels set: test=lifecycle");

                System.out.println("Stopping sandbox");
                sandbox.stop();
                System.out.println("Sandbox stopped");

                System.out.println("Starting sandbox");
                sandbox.start();
                System.out.println("Sandbox started");

                System.out.println("Getting existing sandbox");
                Sandbox fetched = northrays.get(sandbox.getId());
                System.out.println("Got sandbox: " + fetched.getId() + " (state: " + fetched.getState() + ")");

                ListSandboxesQuery query = new ListSandboxesQuery();
                query.setLimit(10);
                query.setLabels(Map.of("env", "dev"));
                query.setStates(List.of(SandboxState.STARTED));
                query.setSort(SandboxListSortField.CREATED_AT);
                query.setOrder(SandboxListSortDirection.DESC);
                for (Sandbox sb : northrays.list(query)) {
                    System.out.println(sb.getId());
                }
            } finally {
                System.out.println("Deleting sandbox");
                sandbox.delete();
                System.out.println("Sandbox deleted");
            }
        }
    }
}

File: examples/python/lifecycle/_async/lifecycle.py
import asyncio

from northrays import AsyncNorthrays, ListSandboxesQuery, SandboxListSortDirection, SandboxListSortField, SandboxState


async def main():
    async with AsyncNorthrays() as northrays:
        print("Creating sandbox")
        sandbox = await northrays.create()
        print("Sandbox created")

        _ = await sandbox.set_labels(
            {
                "public": "true",
            }
        )

        print("Stopping sandbox")
        await northrays.stop(sandbox)
        print("Sandbox stopped")

        print("Starting sandbox")
        await northrays.start(sandbox)
        print("Sandbox started")

        print("Getting existing sandbox")
        existing_sandbox = await northrays.get(sandbox.id)
        print("Get existing sandbox")

        response = await existing_sandbox.process.exec('echo "Hello World from exec!"', cwd="/home/northrays", timeout=10)
        if response.exit_code != 0:
            print(f"Error: {response.exit_code} {response.result}")
        else:
            print(response.result)

        async for sb in northrays.list(
            ListSandboxesQuery(
                limit=10,
                labels={"env": "dev"},
                states=[SandboxState.STARTED],
                sort=SandboxListSortField.CREATEDAT,
                order=SandboxListSortDirection.DESC,
            )
        ):
            print(sb.id)

        print("Removing sandbox")
        await northrays.delete(sandbox)
        print("Sandbox removed")


if __name__ == "__main__":
    asyncio.run(main())
